pkg/descheduler/strategies/duplicates.go

```diff
@@ -67,13 +69,31 @@ func RemoveDuplicatePods(
 		// If any of the existing lists for that first key matches the current pod's list, the current pod is a duplicate.
 		// If not, then we add this pod's list to the list of lists for that key.
 		duplicateKeysMap := map[string][][]string{}
+	PodLoop:
 		for _, pod := range pods {
 			ownerRefList := podutil.OwnerRef(pod)
 			if hasExcludedOwnerRefKind(ownerRefList, strategy) {
 				continue
 			}
+
 			podContainerKeys := make([]string, 0, len(ownerRefList)*len(pod.Spec.Containers))
 			for _, ownerRef := range ownerRefList {
+				var err error
+				replicaCount, ok := ownerReplicationCounter[ownerRef.Name]
+				if !ok {
+					replicaCount, err = podutil.GetPodOwnerReplicationCount(ctx, client, ownerRef)
+					if err != nil {
+						klog.Errorf("Error retreiving replica count for owner %s of %s pod in %s namespace", ownerRef.Name, pod.Name, pod.Namespace)
+						continue
+					}
+					ownerReplicationCounter[ownerRef.Name] = replicaCount
+				}
+				// If required replicas of the managing owner is greater than available nodes, pods will be duplicated
+				// and shouldnt not be considered for eviction
+				if replicaCount > totalNodes {
```

Comment:
Evict duplicated pods only when their owner's `replicaCount` less than `node num` may make descheduler do nothing in some cases. Assume we have 3 nodes and 1 rs and its all 9 replicas lie on the same node, then each node are expected to have 3 replicas for best distribution, so descheduler should evict 6 duplicated pods, but as we only evict duplicated pods when  `replicaCount` less than `node num`, no pod will be evicted.

I think we should calculate the count of duplicated pods that every node expected to have(by `replicaCount` /`node num`), and evicte all duplicated pods above that count. This is exactly what  #263 wants to solve.

What do you think? @damemi  @ingvagabund  @seanmalloy

Reply:
I agree with @lixiang233. The check makes the strategy unnecessary. The idea is to evict the duplicate pods to allow uniform distribution of pods among available nodes. Which will not be true if the check gets merged.

Reply:
@hanumanthan let us know if you have any questions about the change that @ingvagabund and @lixiang233 are requesting. We twill ry to answer any questions you may have. Thanks!

Reply:
I am thinking of approaching this way @lixiang233 @ingvagabund 
Group pods by owner references, and for each `ownerRef` check if `len(duplicated_pods) > replicaCount/totalNodes`.
If yes, then delete the extras. 

Have a doubt - as pod's ownerRef is a list, what if a pod is duplicate for one owner and non-duplicate of another owner - do we evict. 


Capturing the idea in the code below - its not optimized but just for conveying the thought
```
for _, pod := range duplicatePods {
	ownerRefList := podutil.OwnerRef(pod)
	for _, ownerRef := range ownerRefList {
		if podList, ok := ownerRefPodsMap[ownerRef.Name]; ok {
			ownerRefPodsMap[ownerRef.Name] = append(podList, pod)
		} else {
			ownerRefPodsMap[ownerRef.Name] = []*v1.Pod{pod}
		}
	}
}
podsToNotDelete:= make([]*v1.Pod)
for owner, pods := range ownerRefPodsMap {
	if count, ok := ownerReplicationCounter[owner]; ok {
		if count/totalNodes < len(pods) {
			podsToNotDelete := pods[:len(pods)-(count/totalNodes)]
		}
	} 
}

for _, pods := range podsToNotDelete {
	// remove this pod from duplicate pods
}

for _, pod := range duplicatePods {
	if _, err := podEvictor.EvictPod(ctx, pod, node, "RemoveDuplicatePods"); err != nil {
		klog.Errorf("Error evicting pod: (%#v)", err)
		break
	}
}
```

Reply:
> `count/totalNodes < len(pods)`

`count` (as number of replicas of an owner) will always be greater or equal to `len(pods)` (Unless you hit the case when replicas are getting scaled down). On the other side, `len(duplicatePods)` should be equal to the number of replicas - 1 (let's assume a case with a single owner and all pods generated from the same template), i.e. count -1. Which turns the condition into `count/totalNodes < count - 1`. Which turns into `count > totalNodes/(totalNodes-1)`. `totalNodes/(totalNodes-1)` is not greater than 2. So for any RC with 3 or more replicas the condition will always be true. Please correct me if my understanding is incorrect.

> Have a doubt - as pod's ownerRef is a list, what if a pod is duplicate for one owner and non-duplicate of another owner - do we evict.

Yes. The goal is to eliminate as many duplicates as possible (with hope of keeping the workload properly balanced).

Though, you might check how many owners see a pod as duplicated, how many does not. Then based on majority decide if the pod is to be evicted or not.

Reply:
I suggest to do the following:
- get all the duplicated pods
- evict only so many pods so the number of duplicated pods on a node is not greater than `owner.replicas`/`total nodes`
- optionally: Check how many owners see a pod as duplicated, how many does not. Then based on majority decide if the pod is to be evicted or not.

Reply:
> > Have a doubt - as pod's ownerRef is a list, what if a pod is duplicate for one owner and non-duplicate of another owner - do we evict.
>
> Yes. The goal is to eliminate as many duplicates as possible (with hope of keeping the workload properly balanced).

Agree with this, our goal is not to be stringent in deciding what is a duplicate. If it matches in any way, it's a dupe to someone

Reply:
yes count here is `owner.totalReplicas`. 
`len(pods)`  is the length of duplicates pods in the node grouped by owner ref
```
for owner, pods := range ownerRefPodsMap {
		if ownerReplicaCount/totalNodes < len(pods) {
```
> evict only so many pods so the number of duplicated pods on a node is not greater than owner.replicas/total nodes
I was attempting this but my code example is just complicated to convey that. Let me refine it
1. Group duplicated pods by ownerRef - this will be of form `owner - list of duplicated pods`
2. eg) if replica count is 9 and 3 nodes - each node should have max 3 pods. Eliminate anything above - Pods to be eliminated - `len(duplicatedPods grouped by owner) - (replicas/totalNodes)+1` (+1 is because, this node has already one pod which wasnt considered as duplicate)